sympy/assumptions/sathandlers.py

```diff
@@ -2,405 +2,464 @@
-
-    (Add, Implies(AllArgs(Q.integer), Q.integer)),
-    (Add, Implies(ExactlyOneArg(~Q.integer), ~Q.integer)),
-    (Mul, Implies(AllArgs(Q.integer), Q.integer)),
-    (Mul, Implies(ExactlyOneArg(~Q.rational), ~Q.integer)),
-    (Abs, Implies(AllArgs(Q.integer), Q.integer)),
-
-    (Number, CheckOldAssump(Q.negative)),
-    (Number, CheckOldAssump(Q.zero)),
-    (Number, CheckOldAssump(Q.positive)),
-    (Number, CheckOldAssump(Q.nonnegative)),
-    (Number, CheckOldAssump(Q.nonzero)),
-    (Number, CheckOldAssump(Q.nonpositive)),
-    (Number, CheckOldAssump(Q.rational)),
-    (Number, CheckOldAssump(Q.irrational)),
-    (Number, CheckOldAssump(Q.even)),
-    (Number, CheckOldAssump(Q.odd)),
-    (Number, CheckOldAssump(Q.integer)),
-    (Number, CheckOldAssump(Q.imaginary)),
-    # For some reason NumberSymbol does not subclass Number
-    (NumberSymbol, CheckOldAssump(Q.negative)),
-    (NumberSymbol, CheckOldAssump(Q.zero)),
-    (NumberSymbol, CheckOldAssump(Q.positive)),
-    (NumberSymbol, CheckOldAssump(Q.nonnegative)),
-    (NumberSymbol, CheckOldAssump(Q.nonzero)),
-    (NumberSymbol, CheckOldAssump(Q.nonpositive)),
-    (NumberSymbol, CheckOldAssump(Q.rational)),
-    (NumberSymbol, CheckOldAssump(Q.irrational)),
-    (NumberSymbol, CheckOldAssump(Q.imaginary)),
-    (ImaginaryUnit, CheckOldAssump(Q.negative)),
-    (ImaginaryUnit, CheckOldAssump(Q.zero)),
-    (ImaginaryUnit, CheckOldAssump(Q.positive)),
-    (ImaginaryUnit, CheckOldAssump(Q.nonnegative)),
-    (ImaginaryUnit, CheckOldAssump(Q.nonzero)),
-    (ImaginaryUnit, CheckOldAssump(Q.nonpositive)),
-    (ImaginaryUnit, CheckOldAssump(Q.rational)),
-    (ImaginaryUnit, CheckOldAssump(Q.irrational)),
-    (ImaginaryUnit, CheckOldAssump(Q.imaginary))
-    ]:
-
-    register_fact(klass, fact)
+    allargs_integer = AllArgs(x, Q.integer(x))(expr)
+    anyargs_even = AnyArgs(x, Q.even(x))(expr)
+    return Implies(allargs_integer, Equivalent(anyargs_even, Q.even(expr)))
+
+
+### MatMul ###
+
+@class_fact_registry.register(MatMul)
+def _(expr):
+    allargs_square = AllArgs(x, Q.square(x))(expr)
+    allargs_invertible = AllArgs(x, Q.invertible(x))(expr)
+    return Implies(allargs_square, Equivalent(Q.invertible(expr), allargs_invertible))
+
+
+### Pow ###
+
+@class_fact_registry.register(Pow)
+def _(expr):
+    base, exp = expr.base, expr.exp
+    return Implies(Q.real(base) & Q.even(exp) & Q.nonnegative(exp), Q.nonnegative(expr))
+
+@class_fact_registry.register(Pow)
+def _(expr):
+    base, exp = expr.base, expr.exp
+    return Implies(Q.nonnegative(base) & Q.odd(exp) & Q.nonnegative(exp), Q.nonnegative(expr))
+
+@class_fact_registry.register(Pow)
+def _(expr):
+    base, exp = expr.base, expr.exp
+    return Implies(Q.nonpositive(base) & Q.odd(exp) & Q.nonnegative(exp), Q.nonpositive(expr))
+
+@class_fact_registry.register(Pow)
+def _(expr):
+    base, exp = expr.base, expr.exp
+    return Equivalent(Q.zero(expr), Q.zero(base) & Q.positive(exp))
```

Comment:
Would it make sense to combine these like:
```python
@class_fact_registry.register_many(Pow)
def _(expr):
    base, exp = expr.base, expr.exp
    return [
        Implies(Q.real(base) & Q.even(exp) & Q.nonnegative(exp), Q.nonnegative(expr)),
        Implies(Q.nonnegative(base) & Q.odd(exp) & Q.nonnegative(exp), Q.nonnegative(expr)),
        Implies(Q.nonpositive(base) & Q.odd(exp) & Q.nonnegative(exp), Q.nonpositive(expr)),
        Equivalent(Q.zero(expr), Q.zero(base) & Q.positive(exp)),
    ]
```
Maybe it would also be clearer to use `>>` for implies e.g.:
```python
(Q.real(base) & Q.even(exp) & Q.nonnegative(exp))  >>  Q.nonnegative(expr)
```